### DISCUSSION TITLE
Creating a custom lemmatizer

### DISCUSSION THREAD

**Discussion by asier-gutierrez** [explosion/spaCy#8006] [Help: Coding & Implementations] [answered]
Hi, I'm trying to create a new Lemmatizer via config.cfg file.

My nlp definition looks like this:
```
[nlp]
lang = "ca"
pipeline = ["transformer","tok2vec","tagger","morphologizer","parser","ner","lemmatizer"]
...
```

The lemmatizer config looks like this:
```
[components.lemmatizer]
factory = 'ca_lemmatizer'
mode = "rule"
model = null
overwrite = false
```

In the implementation file I have the following code:
```
# Copied from explosion/spaCy/master/spacy/pipeline/lemmatizer.py
@Language.factory(
    "ca_lemmatizer",
    assigns=["token.lemma"],
    default_config={"model": None, "mode": "lookup", "overwrite": False},
    default_score_weights={"lemma_acc": 1.0},
)
def make_lemmatizer(
    nlp: Language, model: Optional[Model], name: str, mode: str, overwrite: bool = False
):
    return CatalanLemmatizer(nlp.vocab, model, name, mode=mode, overwrite=overwrite)

class CatalanLemmatizer(Lemmatizer):
   #overriding
   @classmethod
    def get_lookups_config(cls, mode: str) -> Tuple[List[str], List[str]]:
        if mode == "rule":
            required = ["lemma_lookup", "lemma_rules", "lemma_exc", "lemma_index"]
            return (required, [])
        else:
            return super().get_lookups_config(mode)
    #overriding
    def rule_lemmatize(self, token: Token) -> List[str]:
        ...
```

I load the lookups this way:
```
[initialize.components]

[initialize.components.lemmatizer]
[initialize.components.lemmatizer.lookups]
@misc = "ca_lookups_loader"
data_path = './lemmas/'
```

The function to load lookups is the following:
```
@spacy.registry.misc("ca_lookups_loader")
def load_lookups(data_path):
    from spacy.lookups import Lookups
    lookups = Lookups()
    # "lemma_lookup", "lemma_rules", "lemma_exc", "lemma_index"
    with open(os.path.join(data_path, 'ca_lemma_lookup.json'), 'r') as lemma_lookup, \
         open(os.path.join(data_path, 'ca_lemma_rules.json'), 'r') as lemma_rules, \
         open(os.path.join(data_path, 'ca_lemma_exc.json'), 'r') as lemma_exc, \
         open(os.path.join(data_path, 'ca_lemma_index.json'), 'r') as lemma_index:
        lookups.add_table('lemma_lookup', json.load(lemma_lookup))
        lookups.add_table('lemma_rules', json.load(lemma_rules))
        lookups.add_table('lemma_exc', json.load(lemma_exc))
        lookups.add_table('lemma_index', json.load(lemma_index))
    return lookups

```

I get this error message: 

> ValueError: [E002]  Can't find factory for ''ca_lemmatizer'' for language Catalan (ca). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you'
> re using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components
> ).
> 
> Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentenciz
> er, textcat, textcat_multilabel, ca_lemmatizer, transformer

I don't understand why the second element from the end of the list "Available factories" is the "ca_lemmatizer" if it is not added (supposedly).

My questions are the following:
1) How should I register the component using the config file?
2) Is the table loading right?
3) Is the get_lookups_config method right? Just to clarify, we have created our own tables and we do not want to use the ones from Spacy Lookups library.

Edit: I have even tried with 
```
from spacy.lang.ca import Catalan
@Catalan.factory(
    "lemmatizer",
    assigns=["token.lemma"],
    default_config={"model": None, "mode": "rule", "overwrite": False},
    default_score_weights={"lemma_acc": 1.0},
)
def make_lemmatizer(
    nlp: Language, model: Optional[Model], name: str, mode: str, overwrite: bool = False
):
    return CatalanLemmatizer(nlp.vocab, model, name, mode=mode, overwrite=overwrite)
```
Which results in the same error. I do not get the difference between Language.factory and Catalan.factory except that the name of the lemmatizer is "ca.lemmatizer".

> ValueError: [E002] Can't find factory for ''lemmatizer'' for language Catalan (ca). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're
> using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).
> 
> Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentenciz
> er, textcat, textcat_multilabel, ca.lemmatizer

Many thanks!

**Comment by svlandeg** [comment, marked as answer]
> ValueError: [E002] Can't find factory for ''ca_lemmatizer'' 

There seems to be an issue with single vs. double quotes here, it's looking for the string `'ca_lemmatizer'` instead of just `ca_lemmatizer`. Could you retry with setting your config file to 
```
factory = "ca_lemmatizer"
```
instead?

**Reply by asier-gutierrez** [reply]
It worked, I am not very familiar with the configuration file. I really appreciate your help. Thank you very much!

**Reply by svlandeg** [reply]
No problem, happy to hear it worked!

**Comment by adrianeboyd** [comment]
If an example helps, it should look really similar to the existing `FrenchLemmatizer`:

https://github.com/explosion/spaCy/tree/master/spacy/lang/fr

I think you want to use `Catalan.factory` with the name `lemmatizer`. If it's showing `ca.lemmatizer`, my best guess is something has gone wrong it how it was registered. You should just see `lemmatizer` like you would otherwise, and then if you set `nlp.lang = ca`, you get this lemmatizer instead of the default one.

**Reply by asier-gutierrez** [reply]
I have not been able to use "lemmatizer" it only works with "ca.lemmatizer". I wonder if this can be a bug. The nlp definition of my configuration file is the same as the one you suggested:
```
[nlp]
lang = "ca"
pipeline = ["transformer","tok2vec","tagger","morphologizer","parser","ner","lemmatizer"]
```
